tasks/mrbgems.rake

```diff
@@ -34,20 +31,42 @@ MRuby.each_target do
         f.puts %Q[ */]
         f.puts %Q[]
         f.puts %Q[#include <mruby.h>]
+        f.puts %Q[#include <mruby/proc.h>]
         f.puts %Q[]
         f.write gem_func_decls
-        unless gem_final_calls.empty?
-          f.puts %Q[]
-          f.puts %Q[static void]
-          f.puts %Q[mrb_final_mrbgems(mrb_state *mrb) {]
-          f.write gem_final_calls
-          f.puts %Q[}]
-        end
+        f.puts %Q[]
+        f.puts %Q[static const struct {]
+        f.puts %Q[  void (*init)(mrb_state *);]
+        f.puts %Q[  void (*final)(mrb_state *);]
+        f.puts %Q[} gem_funcs[] = {]
+        f.write gem_func_pairs
+        f.puts %Q[};]
+        f.puts %Q[]
+        f.puts %Q[#define NUM_GEMS ((int)(sizeof(gem_funcs) / sizeof(gem_funcs[0])))]
+        f.puts %Q[]
+        f.puts %Q[static void]
+        f.puts %Q[mrb_final_mrbgems(mrb_state *mrb) {]
+        f.puts %Q[  int ai = mrb_gc_arena_save(mrb);]
+        f.puts %Q[  for (int i = NUM_GEMS - 1; i >= 0; i--) {]
+        f.puts %Q[    gem_funcs[i].final(mrb);]
+        f.puts %Q[    mrb_gc_arena_restore(mrb, ai);]
+        f.puts %Q[  }]
+        f.puts %Q[}]
         f.puts %Q[]
         f.puts %Q[void]
         f.puts %Q[mrb_init_mrbgems(mrb_state *mrb) {]
-        f.write gem_init_calls
-        f.puts %Q[  mrb_state_atexit(mrb, mrb_final_mrbgems);] unless gem_final_calls.empty?
+        f.puts %Q[  int ai = mrb_gc_arena_save(mrb);]
+        f.puts %Q[  for (int i = 0; i < NUM_GEMS; i++) {]
+        f.puts %Q[    gem_funcs[i].init(mrb);]
+        f.puts %Q[    struct REnv *e = mrb_vm_ci_env(mrb->c->cibase);]
+        f.puts %Q[    mrb_vm_ci_env_clear(mrb->c->cibase);]
+        f.puts %Q[    mrb_env_unshare(mrb, e, FALSE);]
```

Comment:
After #5945 merged, we should remove `mrb_env_unshare()` here, do we?

Reply:
Yup, I've been waiting your cue!